ibis/expr/datashape.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import annotations
+
+from typing import Any
+
+from public import public
+
+from ibis.common.grounds import Singleton
+
+
+@public
+class DataShape(Singleton):
+    ndim: int
+    SCALAR: Scalar
+    COLUMNAR: Columnar
+
+    def is_scalar(self) -> bool:
+        return self.ndim == 0
+
+    def is_columnar(self) -> bool:
+        return self.ndim == 1
+
+    def is_tabular(self) -> bool:
+        return self.ndim == 2
+
+    def __lt__(self, other: Any) -> bool:
+        if not isinstance(other, DataShape):
+            return NotImplemented
+        return self.ndim < other.ndim
+
+    def __le__(self, other: Any) -> bool:
+        if not isinstance(other, DataShape):
+            return NotImplemented
+        return self.ndim <= other.ndim
+
+    def __eq__(self, other: Any) -> bool:
```

Comment:
Do we also need an implemenation of `__ne__`?

Reply:
Theoretically no:

> For [__ne__()](https://docs.python.org/3/reference/datamodel.html#object.__ne__), by default it delegates to [__eq__()](https://docs.python.org/3/reference/datamodel.html#object.__eq__) and inverts the result unless it is NotImplemented